File: go/0300_longest_increasing_subsequence.go
func lengthOfLIS(nums []int) int {
	var tails []int
	for i := 0; i < len(nums); i++ {
		tlen := len(tails)
		if tlen == 0 {
			tails = []int{nums[i]}
			continue
		}
		if nums[i] > tails[tlen-1] {
			tails = append(tails, nums[i])
			continue
		}

		start, end, mid := 0, tlen-1, 0
		for start != end {
			mid = start + (end-start)/2
			if nums[i] > tails[mid] {
				start = mid + 1
			} else {
				end = mid
			}
		}
		tails[end] = nums[i]
	}
	return len(tails)
}

File: python/0300_longest_increasing_subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        tails = []
        for num in nums:
            if not tails:
                tails.append(num)
                continue

            pos = self.bisect_right(tails, num)
            if pos == 0:
                tails[0] = num
            elif pos < len(tails):
                if tails[pos - 1] != num:
                    tails[pos] = num
            elif tails[pos - 1] != num:
                tails.append(num)

        return len(tails)

    def bisect_right(self, nums, target):
        lo = 0
        hi = len(nums)
        while lo < hi:
            mid = lo + (hi - lo) // 2
            if nums[mid] <= target:
                lo = mid + 1
            else:
                hi = mid
        return hi